climpred/prediction.py

```diff
@@ -38,19 +47,62 @@ def compute_perfect_model(
         metric (str): `metric` name, see :py:func:`climpred.utils.get_metric_function`.
         comparison (str): `comparison` name, see
                           :py:func:`climpred.utils.get_comparison_function`.
+        dim (str or list): dimension to apply metric over. default: ['member', 'init']
         add_attrs (bool): write climpred compute args to attrs. default: True
 
     Returns:
-        skill (xarray object): skill score.
+        skill (xarray object): skill score with dimension as input without dim.
 
     """
-    supervector_dim = 'svd'
-    metric = get_metric_function(metric, PM_METRICS)
+    if metric in PROBABILISTIC_METRICS:
+        if comparison not in PROBABILISTIC_PM_COMPARISONS:
+            raise ValueError(
+                f'Probabilistic metric {metric} cannot work with comparison \
+                 {comparison}'
+            )
+        stack = False
+        # dim = 'member'
+    elif set(dim) == set(['init', 'member']):
+        supervector_dim = 'svd'
+        stack = True
+    else:
+        if metric in ['pr', 'pearson_r', 'acc']:
+            warnings.warn('ACC doesnt work on dim other than ["init", "member"]')
+        stack = False
+
+    if not stack:
+        supervector_dim = dim
+
     comparison = get_comparison_function(comparison, PM_COMPARISONS)
 
-    forecast, reference = comparison(ds, supervector_dim)
+    forecast, reference = comparison(ds, supervector_dim, stack=stack)
+
+    # in case you want to compute skill over member dim
+    if (forecast.dims != reference.dims) and (metric not in PROBABILISTIC_METRICS):
+        # broadcast when deterministic dim=member
+        forecast, reference = xr.broadcast(forecast, reference)
+        # m2m creates additional forecast_member when over dim member
+        if comparison.__name__ == '_m2m':
+            supervector_dim = 'forecast_member'
+        else:
+            supervector_dim = 'member'
 
-    skill = metric(forecast, reference, dim=supervector_dim, comparison=comparison)
+    metric = get_metric_function(metric, HINDCAST_METRICS)
+
+    skill = metric(
+        forecast, reference, dim=supervector_dim, comparison=comparison, **kwargs
+    )
+
+    # correction for distance based metrics in m2m comparison
+    comparison_name = comparison.__name__[1:]
+    metric_name = metric.__name__[1:]
+    if metric_name in PROBABILISTIC_METRICS and comparison_name == 'm2m':
+        if 'forecast_member' in skill.dims:
+            skill = skill.mean('forecast_member')
```

Comment:
Why is this renamed? Shouldn't it come back as `member`?

Reply:
in the m2m comparison when not stacking (forecast has members, reference does not), we still have each member being the reference in turn. this is reflected by `forecast_member`. maybe I should rename this into `reference_member`